### ISSUE TITLE
Question: Discrepancy between cataloginventory_stock_status and inventory_source_item

### ISSUE THREAD

**Issue by amenk** [AmpersandHQ/magento2-disable-stock-reservation#13]
We introduced the module on our testsystem and truncated inventory_reservations. Not sure it that is the source of the problem.

Then we ordered a product which was having stock = 1. Now in the backend salable quantity in default stock is still 1 while quantity is (correctly) 0.

Do you know if this is expected behavior? Caused by the module or Magento core?

	MariaDB [testing]> select * from cataloginventory_stock_status where product_id = (SELECT entity_id FROM catalog_product_entity WHERE sku = "X6262");
	+------------+------------+----------+--------+--------------+
	| product_id | website_id | stock_id | qty    | stock_status |
	+------------+------------+----------+--------+--------------+
	|      28605 |          0 |        1 | 1.0000 |            1 |
	+------------+------------+----------+--------+--------------+
	1 row in set (0.00 sec)

	MariaDB [testing]> select * from inventory_source_item where sku = X6262";
	+----------------+-------------+-------+----------+--------+
	| source_item_id | source_code | sku   | quantity | status |
	+----------------+-------------+-------+----------+--------+
	|          11389 | default     | X6262 |   0.0000 |      1 |
	+----------------+-------------+-------+----------+--------+
	1 row in set (0.00 sec)

**Comment by amenk** [comment]
I can also reproduce it with an item which was not in stock reservations before.

The salable quantity in the backend does not decrease. 

Do other users of the module have the same problem?

**Comment by amenk** [comment]
I am not sure if it is related, by we disable https://github.com/AmpersandHQ/magento2-disable-stock-reservation/blob/master/src/Plugin/PlaceReservationsForSalesEventPlugin.php#L29 but then in https://github.com/AmpersandHQ/magento2-disable-stock-reservation/blob/master/src/Plugin/SourceDeductionProcessor.php#L136 we call exactly this disabled event.

Is this intentionally?

**Comment by convenient** [comment]
@amenk I intend to write a test to capture this issue against vanilla to see whats happening

**Comment by convenient** [comment]
Hello @amenk 

I have refactored the test suite for better code reuse to target this. The relevant changes from my branch (https://github.com/AmpersandHQ/magento2-disable-stock-reservation/pull/14) can be viewed here https://github.com/AmpersandHQ/magento2-disable-stock-reservation/commit/af1730fe0e4d1c4092fc412d064328bc3f23828e and the corresponding travis build is https://travis-ci.org/github/AmpersandHQ/magento2-disable-stock-reservation/builds/670611900
 
In this test you can see I 
1. create a product with 50 qty
2. verify the values in both stock tables are 50
3. check out with 30 products in the basket (no shipping / invoicing, the order will be at `pending`)
4. verify the values in both tables say 20 remaining
5. verify that adding 30 products to your basket is not possible (because we only have 20 and the requested qty is not available)
6. verify that adding 20 products to your basket works

When I log into the vanilla instance at the end and look at this product I see 20 qty is available 

<img width="1274" alt="Screenshot 2020-04-03 at 15 58 46" src="https://user-images.githubusercontent.com/600190/78376074-b2621180-75c5-11ea-93fb-cb99b04e4c4d.png">

To me this seems that the module is working against vanilla community magento. Am I missing something when trying to replicate your issue? 
- Is your product of a different type maybe? 
- Can you try with `checkmo` locally just to see if it behaves the same as my tests?

Regarding your comment on `PlaceReservationsForSalesEventPlugin`, I have commented out the plugin locally and ran the test suite and the result of this is that there are database writes to `inventory_reservation`. I didn't build this piece of code so would need to look further, but I expect that this plugin disables a lot of logic which includes writing to `inventory_reservation` as well the `placeReservationsForSalesEvent` logic which was then needed to be triggered by ampersand/magento2-disable-stock-reservation at a different point.

Let me know if I can help any further on this, we install this module before the websites go live so it's possible there is some data cleanup issues we have not faced that you may be facing. If you want to do a PR and hook into travis to play with the tests feel free.

**Comment by amenk** [comment]
Thanks a lot for your efforts. I am checking. I would suggest to put a `n98-magerun2 sys:info` into the test pipeline to easily see the Magento version. I cannot sport it. We are oun 2.3.3-p1

**Comment by convenient** [comment]
<img width="276" alt="Screenshot 2020-04-03 at 16 35 22" src="https://user-images.githubusercontent.com/600190/78378635-2356f880-75c9-11ea-9bf4-566e7aeee5dc.png">

@amenk Good idea! For now you can see the versions there. Latest will always build against the latest version of magento and two_three_three is 2.3.3

So we run the suite against 2.3.3 and 2.3.4

**Comment by amenk** [comment]
can we run Travis also against 2.3.3-p1 ?

**Comment by amenk** [comment]
I can reproduce the issue, but only if I do not run cron. If I run cron, both stocks (salable quantity was staying at 50 before the cron run drop to 20 after ordering 30 pieces.

**Comment by convenient** [comment]
@amenk I quickly committed and reverted in my branch another build for 2.3.3-p1

https://travis-ci.org/github/AmpersandHQ/magento2-disable-stock-reservation/builds/670629179
You can see in that build it's using your version and it's gone green

```
Installing magento/project-community-edition (2.3.3-p1): Downloading (100%)
```

FYI we're using https://github.com/AmpersandHQ/travis-vanilla-magento to spin up magento versions in travis by number, you can see the syntax [here](https://github.com/AmpersandHQ/magento2-disable-stock-reservation/pull/14/commits/915aa8937e9931b89222428da9d74fe516ab8589)

**Comment by convenient** [comment]
@amenk just saw your last message. Is my understanding correct here

- Cron enabled = Admin panel and `inventory_source_item` shows the correct value 
- Cron disabled = Admin panel and `inventory_source_item` shows the incorrect value

Is the above the case? If so it's likely stock indexing. In a default magento install (such as what travis is running) the stock indexer will be set to Update On Save rather than Schedule.

If you have your local instance set to Schedule but are not running the cron or are not manually reindexing, would that have caused this issue?

**Comment by amenk** [comment]
Inventory index is update on save ..

I have the crontab always disabled on my dev system, so after I purchase, the saleable quantity in the backend shows the old value (not sure if that is taken from inventory_source_item). As soon as I manually run the cron, the value get corrected. We found the issue on a test system with crontab setup, so this is strange - but there might have been an issue with the crontab there.
On that test system we also truncated `inventory_reservations` after rollout of the module.
I tested again now on that testserver with cron on, and can reproduce the problem. Even after cron is run the saleable quantity stays "1" after I purchased the last stock item.
But: after a full reindex it goes down to 0 - so I believe there is some messup with the index.

On that test server:

Next I set the qty to 1 again for that product.
First it means that the qty becomes 1 and the saleable quantity stays = 0
When I run the cron it still stays 0

when I run `n98-magerun2 index:reindex inventory cataloginventory_stock`

the salable quantity still stays 0 ...

On the test server something is more messed up than locally - locally it works pretty well (cron must be on, but that's okay)


Long story short: 

I saw that our inventory index was update-on-save and cataloginventory_stock was update-on-schedule. I set both to schedule and now the stock correct's after running the cron, might not be a good idea to have both of the indices on different modes.

Still the saleable quantity lags behind the quantity, but it gets fixed on the next minute, which is fine.

Not sure if we should close this issue, maybe we could add a note to the readme? But it seems to be mostly an indexing issue and at the end of the day, it works.

Thanks again for your time!

**Comment by convenient** [comment]
Hey @amenk thanks for all the details

I think a note in the readme to have both the stock indexers on the same mode, and that the indexes need to be triggered for values to be consistent should cover this?

**Comment by amenk** [comment]
Yes

**Comment by convenient** [comment]
@amenk I have updated the README.md [here](https://github.com/AmpersandHQ/magento2-disable-stock-reservation#additional-notes) 

Thanks.
